Add guide on offline update (#409)
foundriesio/docs: source/user-guide/offline-update/offline-update.rst

.. _ug-offline-update:

> Offline Update

Reviewer: I think it probably should be plural for most references in this document "Offline Updates"

Author: But this is a guide on performing a single update :).

Reviewer: lets see what katrina things. it feel wrong when i read it.

Author: By `Offline Update` I mean a concept of updating device offline, so it is a singular noun for me. Of course, user can do many offline updates by applying a concept of Offline Update.

Reviewer 2: I believe the title should be in the plural, as an "Offline Update" is an abstract but countable noun. unless explicitly singular ("Performing an Offline Update"). 
However I would go with "Offline Updates" OR  changing it to a verb and using "Offline Updating"

Author: ok

==============

This section guides you through the steps to update a device offline.

Enabling Offline Update In FoundriesFactory
-------------------------------------------
First of all, you should enable Offline Update in your FoundriesFactory and make sure the target device runs an LmP image that supports an offline update.

> To do so, please, add ``aklite-offline`` to ``PACKAGECONFIG`` of ``aktualizr``, i.e. ``PACKAGECONFIG:append:pn-aktualizr = " aklite-offline"``.
> You can do it in one of the recipes or config files of the :ref:`meta-subscriber-overrides repo <tutorial-customizing-the-platform>`.

Reviewer: I think this would be nice to have it more visually explicit like:
```suggestion
To do so, please, add ``aklite-offline`` to ``PACKAGECONFIG`` of ``aktualizr``. For example::

  # meta-subscriber-overrides.git recipes-sota/aktualizr/aktualizr_git.bbappend
  PACKAGECONFIG:append:pn-aktualizr = " aklite-offline"
```

If you would like to update ``Compose Apps`` along with rootfs (aka ostee), then you must enable :ref:`App Preloading <ref-preloaded-images>` in the factory.

After that, build a LmP image and flash it into a target device or update the device with the image via OTA.

> Also, you should ensure that :ref:`TUF keys are taken offline <ref-offline-keys>`.

Reviewer: Maybe this section could be called "prerequisites" or something so that you don't have this awkward "Also, ..." sentence that kind of feels like an after thought.

This would also help with make the "After that, " sentence feel less awkward. You could just have:
prereqs:
 * a device with an LmP build
 * TUF offline keys

but since TUF offline keys aren't actually required, you could even drop that. We have a going to prod section in the docs, so that's probably where we should keep things.

Reviewer: below you repeat a device needs to have an LMP build. Maybe you don't need this here at all?

Author: > Maybe this section could be called "prerequisites"

I thought about it too.

Author: > since TUF offline keys aren't actually required,

It is required since the `fioctl` starts fetching from v4, otherwise Offline Update won't work.

Obtaining Offline Update Content
--------------------------------
Once Offline Update is enabled, the next step is to download offline update content from the FoundriesFactory to some medium, e.g. a USB drive, which can be attached to a target device.
The offline update content consists of:

1. `TUF metadata`_;
2. `OSTree`_ repo containing a device's rootfs;
3. :ref:`Compose Apps <ref-compose-apps>`.

Use ``fioctl targets offline-update <target-name> <dst> --tag <tag> [--prod]`` command to download the update content.

* ``<target-name>`` - denotes the Target to update a device to
* ``<dst>`` - defines a path to download the update content to
* ``<tag>`` - specifies the Target tag and the tag that the device is on
* ``--prod`` - indicates whether this is an update for a production device and ``<target-name>`` refers to Production Target

Use ``fioctl waves init/complete`` commands to generate and apply :ref:`Production Targets <ref-production-targets>`.

> Also, you must ensure that the target device is a ``production`` device in this case (``--prod`` is specified), see :ref:`Manufacturing Process for Device Registration <ref-factory-registration-ref>` for more details.

Reviewer: this was confusing to read. Maybe we need a "production updates" section or something. Or just state more simply (but at the risk of less actual detail) with a rst note:
```
.. note::

  Production devices require :ref:`production metadata<ref-production-targets>`.
```

Author: Hm, it looks pretty straightforward to me, maybe I am too biased.

- *101* - Restart of dockerd service is required , e.g. ``systemctl restart docker``. After the restart ``aklite-offline run`` must be invoked.
- *0* - OSTree/rootfs installation and Apps start was successful, update completed. Only ``aklite-offline run`` can return this code.

If ``aklite-offline run`` returns *100* , then an update has not been successful and rollback has happened. In this case you should trigger reboot followed by another ``aklite-offline run`` to complete the rollback.

Configuration Details
~~~~~~~~~~~~~~~~~~~~~

The minimum required configuration is

.. prompt:: text

    [provision]
    primary_ecu_hardware_id = <>

The command can digest the default device config that consists of:

1. ``*toml`` files added into LmP during bitbaking (usually just /usr/lib/sota/conf.d/40-hardware-id.toml);
2. sota.toml generated by ``lmp-device-register``.

If a device needs to support both types of update, offline and online, then the configuration should be shared with ``aktualizr-lite``.

If you intend to update a device only offline, then it’s suffice to specify ``[provision].primary_ecu_hardware_id`` in the configuration.

> Normally, each LmP image includes it by default in this file ``/usr/lib/sota/conf.d/40-hardware-id.toml``.

Reviewer: I'd start this section with this sentence. I started to panic about doing all this work (and started looking to verify) *then* i hit this sentence.

Author: makes sense.

.. _ug-offline-update:

Offline Update
==============

> This section guides you through the steps to update a device offline.

Reviewer 2: ```suggestion
This section guides you through the steps of updating a device offline.
```

Author: ok, this is copy-paste of https://docs.foundries.io/latest/user-guide/container-preloading/container-preloading.html?guides#container-preloading :)

Reviewer 2: I'll make a note to update that as well!

Prerequisites
-------------

1. Enable Offline Update support, i.e. add ``aklite-offline`` to ``PACKAGECONFIG`` of ``aktualizr``. For example:

    .. prompt:: text

        cat meta-subscriber-overrides.git/recipes-sota/aktualizr/aktualizr_%.bbappend

>         PACKAGECONFIG:append:pn-aktualizr = " aklite-offline"

Reviewer 3: Here I would recommend using:

`PACKAGECONFIG:append = "aklite-offline"`

pn-aktualizr can be used in a global config, here we can simply just extend PACKAGECONFIG instead.

Author: Great, thanks!

> 2. Enable :ref:`App Preloading <ref-preloaded-images>`, if you would like to update ``Compose Apps`` along with rootfs (aka ostree).

Reviewer 2: ```suggestion
2. Enable :ref:`App Preloading <ref-preloaded-images>` if you would like to update ``Compose Apps`` along with rootfs (aka ostree).
```

Performing Offline Update
-------------------------
Before doing offline update make the offline update content accessible on a device, e.g. attach and mount the USB drive.

Use the ``aklite-offline`` CLI utility to perform an offline update:

1. Run ``aklite-offline install [--config <config dir or file>] --src-dir <path to offline update content>``.

2. Run one of the post installation actions depending on the ``aklite-offline`` install result:

    a. trigger reboot to perform a rollback if the update failed;
    b. reboot device and invoke ``aklite-offline run [--config <config dir or file>]`` to finalize an ostree installation and start Apps if both ostree/rootfs and Apps are updated;
    c. restart the Docker Engine (e.g. ``systemctl restart docker``) and invoke ``aklite-offline run [--config <config dir or file>]``  if just Apps are updated.

Usage Details
-------------
The CLI utility supports two commands:

1. ``aklite-offline install [--config <config file/dir>] --src-dir <update-content-dir>``
2. ``aklite-offline run [--config <config file/dir>]``

.. prompt:: text

>     --config -  Path to a directory that contains one of more *.toml configuration snippets or a path to a *.toml file. It may be omitted at all so the command collects config from the snippets found in the default directories/files, as aktualizr-lite does:

Reviewer 4: maybe?

```suggestion
    ``--config`` -  Path to a directory that contains one of more ``*.toml`` configuration snippets or a path to a ``*.toml`` file. It may be omitted at all so the command collects config from the snippets found in the default directories/files, as ``aktualizr-lite`` does:
```

    /usr/lib/sota/conf.d
    /var/sota/sota.toml
    /etc/sota/conf.d/

>     --src-dir - Path to a directory that contains update content downloaded by ``fioctl targets offline-update`` command.

Reviewer 4: ```suggestion
    ``--src-dir`` - Path to a directory that contains update content downloaded by ``fioctl targets offline-update`` command.
```

Post Install Actions
~~~~~~~~~~~~~~~~~~~~

> The command sets the following exit codes ($?) that instructs which of the post install actions you should perform:

Reviewer 4: ```suggestion
The command sets the following exit codes (``echo $?``) that instruct which of the post-install actions you should perform:
```

- *100* - Reboot is required to complete installation. After reboot ``aklite-offline run`` must be invoked.
- *101* - Restart of dockerd service is required , e.g. ``systemctl restart docker``. After the restart ``aklite-offline run`` must be invoked.
- *0* - OSTree/rootfs installation and Apps start was successful, update completed. Only ``aklite-offline run`` can return this code.

If ``aklite-offline run`` returns *100* , then an update has not been successful and rollback has happened. In this case you should trigger reboot followed by another ``aklite-offline run`` to complete the rollback.

Configuration Details
~~~~~~~~~~~~~~~~~~~~~

> The minimum required configuration is

Reviewer 2: ```suggestion
The minimum required configuration is:
```

.. prompt:: text

    [provision]
    primary_ecu_hardware_id = <>

> The command can digest the default device config that consists of:

Reviewer 2: ```suggestion
The command can digest the default device config consisting of:
```

> 1. ``*toml`` files added into LmP during bitbaking (usually just /usr/lib/sota/conf.d/40-hardware-id.toml);

Reviewer 4: ```suggestion
1. ``*toml`` files added into LmP during bitbaking (usually just ``/usr/lib/sota/conf.d/40-hardware-id.toml``);
```

> 2. sota.toml generated by ``lmp-device-register``.

Reviewer 4: ```suggestion
2. ``sota.toml`` generated by ``lmp-device-register``.
```

If a device needs to support both types of update, offline and online, then the configuration should be shared with ``aktualizr-lite``.

Normally, each LmP image includes a configuration file ``/usr/lib/sota/conf.d/40-hardware-id.toml`` which defines a hardware ID.
Therefore, by default, a LmP image includes the minimum required configuration, hence NO device registration is required for ``aklite-offline`` to work.

> If you register a device and ``sota.toml`` is generated, then the offline update command can either work alone or co-work with ``aktualizr-lite``.

Reviewer 2: ```suggestion
If you register a device and ``sota.toml`` is generated, then the offline update command can either work alone or along with ``aktualizr-lite``.
```

> In the later case, you must stop the ``aktualizr-daemon`` before running the offline update command.

Reviewer 3: Maybe 'you must stop the aktualizr-lite systemd service' instead?

Author: Makes sense. Updated.

.. _ug-offline-update:

Offline Update
==============

This section guides you through the steps to update a device offline.

Prerequisites
-------------

1. Enable Offline Update support.

    Add ``aklite-offline`` to ``PACKAGECONFIG`` of ``aktualizr``, for example:

    ``PACKAGECONFIG:append:pn-aktualizr = " aklite-offline"``.

>     You can do it in one of the recipes or config files of the :ref:`meta-subscriber-overrides repo <tutorial-customizing-the-platform>`.

Reviewer: i still think you should give a concrete filename and contents here.

Author: But it can be any file/recipe.

Author: Maybe we can have kind of best practice that we recommend to our customer in such the case, @ricardosalveti what do you think?
If so, I can state it as `for example:`.

Reviewer: it can be any file and for people that understand that - they'll gloss over this. for users that don't know that- we don't really need to make them read another 5 minutes and feel uncertain.
